### ISSUE TITLE
error TS2612 Property will overwrite base property not raised if "strictNullCheck" is true

### ISSUE THREAD

**Issue by amodiopescefaro** [microsoft/TypeScript#48667]
# Bug Report

If you enable [useDefineForClassFields](https://www.typescriptlang.org/tsconfig#useDefineForClassFields) then a new error is possible TS26126

If you have code that trigger TS2612 Property 'xxx' will overwrite the base propert in 'yyy'. It this is intentional, add an initializer. Othertwise, add a "declare" modifier or remove the redundant declaration.

The error do not appear if you enable strictNullCheck.

Without strictNullCheck you'll see this error.

If you put strict=true or strictNullCheck=true the error is not reported anymore

### 🔎 Search Terms

TS2612
strictNullCheck

### 🕗 Version & Regression Information

Noticed this on typescript 4.6.3
Tested on typescript@nightly 4.7.0-dev.20220408 and it behave the same (no error on strict mode)
Tested on 4.3.5 and it behave the same (no error on strict mode)

TS2612 was introduced in 4.3 so previous version do not make sense

### ⏯ Playground Link

This link contain a minimal code to show the problem but you need to enable/disable strictNullCheck and [useDefineForClassFields](https://www.typescriptlang.org/tsconfig#useDefineForClassFields) to see it in action

[Playground link with relevant code](https://www.typescriptlang.org/play?ssl=33&ssc=2&pln=10&pc=1#code/PTAEHUFMBsGMHsC2lQBd5oBYoCoE8AHSAZVgCcBLA1UABWgEM8BzM+AVwDsATAGiwoBnUENANQAd0gAjQRVSQAUCEmYKsTKGYUAbpGF4OY0BoadYKdJMoL+gzAzIoz3UNEiPOofEVKVqAHSKymAAmkYI7NCuqGqcANag8ABmIjQUXrFOKBJMggBcISGgoAC0oACCbvCwDKgU8JkY7p7ehCTkVDQS2E6gnPCxGcwmZqDSTgzxxWWVoASMFmgYkAAeRJTInN3ymj4d-jSCeNsMq-wuoPaOltigAKoASgAywhK7SbGQZIIz5VWCFzSeCrZagNYbChbHaxUDcCjJZLfSDbExIAgUdxkUBIursJzCFJtXydajBVDtUAAIQYghQAF5QABvRQlErSWmQfJXVCUTjMADcigAvkLFBSiKAACLfXSQVyMml00AAMmZrLZ3Flem43MEvOGQuFwVgjEEwiqLLZ8zYBG5SsgYutCE4+rI7Fg6DIAAoCLaAPz2zkASnV1utQ0EAT98AIoEZMbj-v9zPGnO5ACIOXSM6BRRqSsbjYpTbThFTwasFDwLWG2YnuTLKDqnWyXW6PV7fbbG9r5aGreGSoJ2BtvcGhUOSpHo7b4zbY5O2UXgiocM0PGQvIh4H0GMD2DQvm4zMx2AxmJB+Kb1Il9-A9CIvBmAKKrBiIBb6XO70AZ8AOKgADkwgAHKQBIGZBCoADyXxkO8dL8JeRyoI4CiuNIeCgE4O46MMWD6CgCCINCwiXMeEi7tEIgGBw2ILEwrAcDwQQlEAA)

### 💻 Code

```ts
type Base = {
    base: string;
};

type Derived = Base & {
    derived: string;
}

class A {
    prop: Base;

    constructor(prop?: Base) {
        this.prop = prop ?? { base: "base" };
    }
}

class B extends A {
    // This should trigger TS2612 but does not do it with strictNullCheck enabled
    prop: Derived;

    constructor(prop: Derived) {
        super();
        this.prop = prop;
    }
}

```

### 🙁 Actual behavior

When compiled with:
* target - esnext
* useDefineConfigForClass - true
* strictNullCheck - true

the code does not raise any error

### 🙂 Expected behavior

It should raise the TS2612 error

**Comment by sandersn** [comment]
Specifically, `B.prop` has an initialiser: in B's constructor, `this.prop = prop`. This means that the compiler thinks your overwrite of the base property is intentional.

**Comment by iamgabrielsoft** [comment]
or you could change it to es2015, I think it is by design for esnext
